src/routing_membrane.rs

```diff
@@ -557,7 +561,12 @@ impl<F> RoutingMembrane<F> where F: Interface {
                         message.clone(), request.clone()),
                     MessageType::GetDataResponse(ref response) =>
                         self.handle_get_data_response(message_wrap, message.clone(), response.clone()),
-                    MessageType::PutData(ref data) => self.handle_put_data(message_wrap, message.clone(), data.clone()),
+                    MessageType::PutData(ref data) => {
+                        match message.source.actual_source() {
+                            Address::Node(name) => self.handle_group_put_data(message_wrap, message.clone(), data.clone(), name),
+                            Address::Client(name) => self.handle_client_put_data(message_wrap, message.clone(), data.clone(), name),
+                        }
+                    },
```

Comment:
I think it is not correct to use `match message.source.actual_source()` here to determine whether the message came from a group. It is because source being a node does not imply it came from a group. I think better would be to use `message.from_authority().is_group()` (it has been recently added, so you'll need to pull from master).